File: software-engineering/lowest-common-ancestor/golang/main_test.go
package main

import "testing"

func TestLCA(t *testing.T) {
	//build tree
	root := TreeNode{Val: 6}
	root.Left = &TreeNode{Val: 2}
	root.Right = &TreeNode{Val: 8}
	root.Left.Left = &TreeNode{Val: 0}
	root.Left.Right = &TreeNode{Val: 4}
	root.Left.Right.Left = &TreeNode{Val: 3}
	root.Left.Right.Right = &TreeNode{Val: 5}
	root.Right.Left = &TreeNode{Val: 7}
	root.Right.Right = &TreeNode{Val: 9}

	//make testing table
	table := []struct {
		x *TreeNode
		y *TreeNode
		n int
	}{
		//LCA root
		{root.Left, root.Right, 6},
		//LCA left node
		{root.Left.Right.Left, root.Left, 2},
		//LCA single node
		{root.Left.Right.Right, root.Left.Right.Right, 5},
		//LCA right node
		{root.Right.Right, root.Right.Left, 8},
		//LCA deepest branches
		{root.Left.Right.Left, root.Left.Right.Right, 4},
	}

	//try each case
	for _, row := range table {
		result := lowestCommonAncestor(&root, row.x, row.y)
		if result.Val != row.n {
			t.Errorf("LCA was incorrect, got: %d, want: %d.", result.Val, row.n)
		}
	}
}

func TestDAG(t *testing.T) {
	//Current implementation cannot handle nodes with more than 2 offspring.
	// TODO more code will have to be added.
}

File: software-engineering/lowest-common-ancestor/cpp/main_test.cpp
#include "main.cpp"
#include "catch_amalgamated.hpp"
#include "catch_amalgamated.cpp"

TEST_CASE() {
    //build tree
    Node * root = addNode(6);
    root->left = addNode(2);
    root->right = addNode(8);
    root->left->left = addNode(0);
    root->left->right = addNode(4);
    root->left->right->left = addNode(3);
    root->left->right->right = addNode(5);
    root->right->left = addNode(7);
    root->right->right = addNode(9);

    //test cases
    //LCA root
    REQUIRE( lowestCommonAncestor(root, root->left, root->right)->val == 6 );
    //LCA left node
    REQUIRE( lowestCommonAncestor(root, root->left->right->left, root->left)->val == 2 );
    //LCA single node
    REQUIRE( lowestCommonAncestor(root, root->left->right->right, root->left->right->right)->val == 5 );
    //LCA right node
    REQUIRE( lowestCommonAncestor(root, root->right->right, root->right->left)->val == 8 );
    //LCA deepest branches
    REQUIRE( lowestCommonAncestor(root, root->left->right->left, root->left->right->right)->val == 4 );
}